samples/bluetooth/extended_adv/scanner/src/main.c

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright (c) 2024 Croxel, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+
+#define NAME_LEN 30
+
+static struct bt_conn *default_conn;
+bt_addr_le_t ext_addr;
+
+enum bt_sample_scan_evt {
+	BT_SAMPLE_EVT_EXT_ADV_FOUND = BIT(0),
+	BT_SAMPLE_EVT_CONNECTED = BIT(1),
+	BT_SAMPLE_EVT_DISCONNECTED = BIT(2),
+	BT_SAMPLE_EVT_SCAN_DUE = BIT(3),
+};
+
+enum bt_sample_scan_st {
+	BT_SAMPLE_ST_SCANNING,
+	BT_SAMPLE_ST_CONNECTING,
+	BT_SAMPLE_ST_CONNECTED,
+	BT_SAMPLE_ST_COOLDOWN,
+};
+
+static volatile enum bt_sample_scan_st app_st = BT_SAMPLE_ST_SCANNING;
+
+static struct k_poll_signal poll_sig = K_POLL_SIGNAL_INITIALIZER(poll_sig);
+static struct k_poll_event poll_evt = K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+						K_POLL_MODE_NOTIFY_ONLY, &poll_sig);
+
+static void connected_cb(struct bt_conn *conn, uint8_t err)
+{
+	printk("Connected (err 0x%02X)\n", err);
+
+	if (err) {
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+		return;
+	}
+
+	k_poll_signal_raise(poll_evt.signal, BT_SAMPLE_EVT_CONNECTED);
+}
+
+static void disconnected_cb(struct bt_conn *conn, uint8_t reason)
+{
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+
+	printk("Disconnected (reason 0x%02X)\n", reason);
+}
+
+static void recycled_cb(void)
+{
+	printk("Connection object available from previous conn. Disconnect is complete!\n");
+	k_poll_signal_raise(poll_evt.signal, BT_SAMPLE_EVT_DISCONNECTED);
```

Comment:
I would avoid using the `result` field of the signal. It's not safe to store "which signal is raised", like we do here. If a `connected` and `disconnected` event follow rapidly, the `result` field of the signal will be overwritten before the handler thread has read it.

Reply:
They can share a single polling signal. But the actual events must be encoded in something that accurately represents the work-to-be-done.

Reply:
I see - that's neat. I think that'll be simpler/easier to refactor. Thanks!